controllers/quay/quayregistry_status_controller.go

```diff
@@ -0,0 +1,319 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	routev1 "github.com/openshift/api/route/v1"
+
+	"github.com/go-logr/logr"
+	ocsv1a1 "github.com/kube-object-storage/lib-bucket-provisioner/pkg/apis/objectbucket.io/v1alpha1"
+	qv1 "github.com/quay/quay-operator/apis/quay/v1"
+)
+
+const (
+	// all objects that are deployed as part of any of our components must contain this
+	// label. it is used when evaluating the conditions for all the components.
+	quayComponentLabel = "quay-component"
+)
+
+// ConditionFetcher is a function capable of returning a map of conditions indexed by component
+// name.
+type ConditionFetcher func(context.Context, qv1.QuayRegistry) (map[string][]qv1.Condition, error)
+
+// QuayRegistryStatusReconciler updates status for QuayRegistry components. This status Reconciler
+// has to live in a different controller to avoid having to have a resync period in the main Quay
+// reconciler as it generally takes longer and also runs a dabase migration everytime it is called.
+type QuayRegistryStatusReconciler struct {
+	Client client.Client
+	Log    logr.Logger
+	Mtx    *sync.Mutex
+}
+
+// NewQuayRegistryStatusReconciler returns a new QuayRegistryStatusController configured to use
+// the provided client.
+func NewQuayRegistryStatusReconciler(cli client.Client) *QuayRegistryStatusReconciler {
+	return &QuayRegistryStatusReconciler{
+		Client: cli,
+		Log:    ctrl.Log.WithName("controllers").WithName("QuayRegistryStatus"),
+	}
+}
+
+// SetupWithManager sets up provided manager as the source of events for this reconciler.
+func (q *QuayRegistryStatusReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).For(&qv1.QuayRegistry{}).Complete(q)
+}
+
+// Reconcile is called for reconcile status for a given QuayRegistry components. This function
+// always rescheduled the same event on return, that makes this to run from time to time.
+func (q *QuayRegistryStatusReconciler) Reconcile(
```

Comment:
Is it a common practice in operators to create a controller to report status of watched resources? Is there no way to have the QuayRegistry controller watch over its owned resources instead of just its own CR?

Reply:
No, it is not a common practice (this is more of a bad practice). The QuayRegistry controller acts only when something changed in a QuayRegistry object and it does a few things that I still don't fully understand (I have been informed that it creates pods for running database upgrades, for instance). To have the QuayRegistry reconcile to happen more frequently might cause other issues and I don't feel comfortable in making such a call this close to feature freeze, hence this hacky approach to the problem. i.e. we need to review the QuayRegistry reconciliation logic before using it more frequently.

Reply:
I have been looking into this and, for example, every time the Reconcile loop of the QuayRegistry controller runs a new Secret is created (named `registry-quay-config-secret-*`) and this secret is then mounted inside a pod causing a full blown redeploy (the old secret stays there, for whatever reason it is not deleted). I imagine how many other surprises we gonna find out if we decide to run this Reconcile loop frequently.